### PR TITLE
Simplify arity errors, rewording spread errors

### LINKED ISSUE #28010: Expected at least x arguments, but got x or more
**TypeScript Version:**  3.2.0-dev.20181019 and 3.1.3

**Search Terms:** function arguments count, spread, rest
**Code**

```ts
function foo(a: number, ...r: number[]): void { };

const a = [1];

foo(...a, 2); // Expected at least 1 arguments, but got 1 or more.
foo(...a, 2, 3); // Expected at least 1 arguments, but got 2 or more.
```

**Expected behavior:**
No errors
"at least 1" == "1 or more"

**Actual behavior:**
"at least 1" != "1 or more"
Non logical type error.

**Playground Link:** [here](https://www.typescriptlang.org/play/index.html#src=function%20foo(a%3A%20number%2C%20...r%3A%20number%5B%5D)%3A%20void%20%7B%20%7D%3B%0A%0Aconst%20a%20%3D%20%5B1%5D%3B%0A%0Afoo(...a%2C%202)%3B%20%2F%2F%20Expected%20at%20least%201%20arguments%2C%20but%20got%201%20or%20more.%0Afoo(...a%2C%202%2C%203)%3B%20%2F%2F%20Expected%20at%20least%201%20arguments%2C%20but%20got%202%20or%20more.)

**Related Issues:** 
#26350

### LINKED ISSUE #42891: Confusing error messages: expected N arguments
# Bug Report

### 🔎 Search Terms

error, arguments, overloads

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?noImplicitAny=false&jsx=0&module=1&ts=4.2.0-beta#code/JYWwDg9gTgLgBAMwM5wLxygUwI4FdhYAUARMsQJQDcAUKJLHEjFgIYhoY75HFOsgUadaPFwxgAGw5Y8BTCTGTBtcCLgAvCcABG0rnJKady6gBNMAYwkssiXADsL4iPcZgtMQiBYwLAC0woAC44ACVMAHMAUQAPMAAaOG8wMECQwi17TBC+YHsI8jQAPjgWewBPchzmTDYAOgAVKDKkBGgQGjNLa1sEBydgFzgYTCZCGJCy8sTy9MLUEtz8qsZmPIi4AB84B3MEPMxTGnMrG0w7R2dXACskFyaWtqgQQhgQmHLUiARh0ZgVvi1ECNZr2VrtY7dM4XAZDUzAJBgcaTCoAtb5TonHrnCwuJhwPJgMTVKDrSGnWy4sHwCBiIlvVakjHUahU-FgYCpTLndCKCR1MBQCAgBHABDlQiA+ocrkHKgslhIcqOGFXOCCiAWUZIABiknkCEVMAAshBzCFtBAIBJavZCgBvahwZ2lADuLGA8BlmG5hCdLoDyDqFlYI3CLFMAGUamxCISxOR4v6A866mnCAB6AA6dQi6gAJBm6iMxvH-nAAPxwADaRm0wdDmAA4g51JzCOQALrJlMBkLZuraKCF4t-OP2emFKu1rT1kO1EYAISFMC0ABFLMLBdqO93e-uXSFq53Ez3A0bTeYz-uq4iPJn4hWs1An-YM4kYsU4DE6jb8jA-DgEoAEZK1+MYAClIwAeQAOQFGwkHkGJEzgDsvxQuAQl2TB9iyUxyGvXsQlue5QXBZ5Xj+U8D0VZULFVQZ7AAKjQwkHSIlMnjdD14EINl4E-b4CQnQpymAH1TDgeFEXGeUDwDcTJLgYgs3sTs1OIGgDwAXyTA80zqTMczzEcS08WkYEnMCZx0BsF2bNskS7Ti+zgAchzMsdLOs6c63snxMGXCBV2AABhLcsCQJBd1c-cjxPfT9yDedAoAdVJEZo34QgfITHt5R0oA)

### 💻 Code

Reduced from https://github.com/elibarzilay/process-tracing/blob/59a790c839258ee737481edad0a614f094a1d68c/index.js

```ts
import fs = require("fs");
import stream = require("stream");
import util = require("util");
import zlib = require("zlib");

declare function split(matcher: RegExp, mapper: (line: string) => any): stream.Transform;

declare function test(x: any, y: () => string): string | undefined;
declare function jsonTransform(t: typeof test): stream.Transform;
declare function disp(x: any): string;

declare const input: string;
declare const output: string;

const pipeline = util.promisify(stream.pipeline);

async function processFile(fastMode: boolean) {
    await pipeline(
        fs.createReadStream(input),
        ...(/\.gz$/.test(input) ? [zlib.createGunzip()]
            : /\.br$/.test(input) ? [zlib.createBrotliDecompress()]
                : []),
        fastMode
            ? split(/,?\r?\n/, x => x.length > 1 ? test(JSON.parse(x), () => x) : undefined)
            : jsonTransform(test),
        async function* (inp) {
            for await (const x of inp) yield disp(x);
            yield "\n]\n";
        },
        ...(/\.gz$/.test(output) ? [zlib.createGzip()]
            : /\.br$/.test(output) ? [zlib.createBrotliCompress()]
                : []),
        fs.createWriteStream(output)
    );
}
```

There are several weird things going on here:

1. `fs.createWriteStream(output)` is squiggled and the error produced is `Expected at least 1 arguments, but got 6 or more.`  There are no elaborations.  I have no idea what's actually wrong.
2. As @minestarks points out, it's not clear how any numbers could be substituted into [this error template](https://github.com/microsoft/TypeScript/blob/9950b6e596c7f847876271fe7d0ca488dca115c6/src/compiler/diagnosticMessages.json#L2393) to make the statement sensible.
3. In 4.1.5 or 4.2.0-dev, if you comment out the `async function *` argument, the new error is `No overload expects 4 arguments, but overloads do exist that expect either 4 or Infinity arguments.`  The phrasing "4 or Infinity" is strange, the value 4 is both expected and not expected, and it's not clear where 4 actually comes from (maybe it assumes each spread is empty?).
4. In 4.2.0-dev.20210219, the squiggle range is reduced to only _some_ of the arguments and the error is `Expected at least 1 arguments, but got 4 or more.`

### PR DESCRIPTION
**Pull request by sandersn**
Arity errors can produce some really strange messages. When I went back and re-learned the arity error rules, I decided the biggest problem is that spread errors are actually a special case, different from other arity errors: spreads of non-tuples are only allowed when the target parameter is a rest:

```ts
declare function fn(x, ...args)
fn(...[1,2,3]) // error
fn(1, ...[2,3]) // ok
```

This is true regardless of type (module having a tuple type, eg `[1,2,3] as const`), which is why I didn't annotate any in the example above.

Previously, the error message for a mismatched spread was similar to the normal error: "Expected 2 arguments, but got at least *n*." This didn't indicate the true problem at all. Add to this that *n* was wrong for calls with multiple spreads, and that the error span was wrong too. It used the span for "too many arguments", which is nonsensical.

This PR:

- Issues a new error message for disallowed spreads.
- Doesn't mention *any* parameter counts, since they're not relevant.
- Uses the spread as the error span.
- Does **not** attempt to add new call resolution rules for non-tuple spreads. (I have a couple of ideas here, but I'm not sure how widely applicable or feasible they are.)

It also restructures almost all the code in getArgumentArityError and renames most variables. The function had become baroque even for its previous requirements, but simplifying the spread error message allows it to read in a much straighter line than before. I'd recommend viewing this in side-by-side diff view on a wide-ish monitor.

Fixes #42891 
Fixes #28010
Addresses #42508 but does not fix it. I added its test case for this PR.

I'm going to experiment with adding non-tuple spread call resolution rules next.

### CODE REVIEW

**Comment by sandersn** [comment]
@amcasey @andrewbranch I added you to the review because you usually care about error messages, and because you were involved in the precipitating bugs. You don't necessarily need to review the code, but of course you are welcome.

**Comment by DanielRosenwasser** [comment]
> A spread must either have a tuple type or be passed to a rest parameter.

So this is true today - however, I can totally imagine a future where the following could be allowed:

```ts
function foo<T extends [string, number, boolean]>(stuff: T) {
    bar(...stuff)
}

function bar(x: string, y: number, z: boolean) {
}
```

and so I guess the core of this change depends on us always rejecting this code.

**Comment by sandersn** [comment]
The core of this change is to stop reporting confusing param/arg indices where we should be reporting an arbitrary limitation on spreads. I'm not sure I've explained that limitation well in the error. And the explanation may need to change if the limitation loosens in the future.

**Comment by sandersn** [comment]
@typescript-bot run dt

(Unrelated to this PR, I want to see if my DT on-demand fix worked)

**Comment by andrewbranch** [comment]
> Addresses #42508 but does not fix it. I added its test case for this PR.
>
> I'm going to experiment with adding non-tuple spread call resolution rules next.

To clarify, are you saying that #42508 should be a wontfix / design limitation, or are you saying you still hope to fix it soon, but in a separate PR?

**Comment by sandersn** [comment]
I hope to fix #42508 soon. I'm going to try treating `[number, number] | [number, number, number]` like `[number, number, number?]` but it may not work out.

**Review by andrewbranch — APPROVED** [review]
I think this closes https://github.com/microsoft/TypeScript/issues/28010 since the decision was just to clean up the nonsense error, which this does.

**Review by amcasey — COMMENTED** [review]
> And the explanation may need to change if the limitation loosens in the future.

It would be easy to forget to update this resource string, so it would be nice if we could come up with text that would continue to work if Daniel's plausible change is made.

Otherwise, seems sensible.  I only skimmed the product changes.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -28559,88 +28559,67 @@ namespace ts {
...
-            let min = Number.POSITIVE_INFINITY;
-            let max = Number.NEGATIVE_INFINITY;
-            let belowArgCount = Number.NEGATIVE_INFINITY;
-            let aboveArgCount = Number.POSITIVE_INFINITY;
+            const spreadIndex = getSpreadArgumentIndex(args);
+            if (spreadIndex > -1) {
+                return createDiagnosticForNode(args[spreadIndex], Diagnostics.A_spread_must_either_have_a_tuple_type_or_be_passed_to_a_rest_parameter);
+            }
+            let min = Number.POSITIVE_INFINITY; // smallest parameter count
+            let max = Number.NEGATIVE_INFINITY; // larger parameter count
```
[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply src/compiler/checker.ts:28567]: largerst

**Comment by sandersn** [comment]
I'll line up a time to talk with @DanielRosenwasser and see if we can come up with a message that is forward-compatible, although I think the current state is already close to fit for that purpose.

**Comment by sandersn** [comment]
A draft of a couple more expansions for spread arguments is up at #43882. However, after implementing the prototype, I'm even less confident that they're correct.

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -2386,14 +2386,10 @@
         "category": "Error",
         "code": 2555
     },
-    "Expected {0} arguments, but got {1} or more.": {
+    "A spread must either have a tuple type or be passed to a rest parameter.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json:2389]:
```suggestion
    "A spread argument must either have a tuple type or must be passed to a rest parameter.": {
```

Comment (andrewbranch) [reply src/compiler/diagnosticMessages.json:2389]: 👍 on “argument,” 👎 on the second “must”

**Pull request merged**